### PR TITLE
Assignment 3 Michael Moran (mmoran314) and Ben Fristad (BenFristadSchool)

### PR DESCRIPTION
**Pull request by mmoran314**
Filtered Chuck Norris Jokes,
Added Unit Tests
Added JokeOutput Class

### CODE REVIEW

**Review by Nolan-Posey — COMMENTED** [review]
Assignment
For this assignment do the following:

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall you guys did a great job! The program compiles and runs as intended. There are only two things you missed:

1.)  The code coverage for the Jester class is under 90% (The Fine Code Coverage extension I'm using is say that it is at 55.5%). I would recommend making some mock classes for JokeOutput and JokeService inside of your test class to use inside of a test jester that way you have more control over the tests. Also I would recommend splitting the CreateJester_Failure() test into two tests, one that expects a null argument exception when  JokeOutput is null and another when Joke service is null.

2.) In the TestWriteJokeConsoleOutput() test you guys didn't use the generics version of Assert.AreEqual (Assert.AreEqual<T>). I would recommend using Assert.AreEqual<string> since it appears you are trying to compare the string intended to be output and the string that the Jester class is outputting.

**Comment by jacobnewcomb** [comment]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Didn't seem to use generic version of Assert.AreEqual(), but everything else looks great and seems to work well!

**Comment by mmoran314** [comment]
>
> Didn't seem to use generic version of Assert.AreEqual(), but everything else looks great and seems to work well!

added <string> to tests.

**Comment by mmoran314** [comment]
> 1.) The code coverage for the Jester class is under 90% (The Fine Code Coverage extension I'm using is saying that coverage for the Jester class is at 55.5%). I would recommend making some mock classes for JokeOutput and JokeService inside of your test class to use inside of a test jester that way you have more control over the tests. Also I would recommend splitting the CreateJester_Failure() test into two tests, one that expects a null argument exception when JokeOutput is null and another when Joke service is null.
> 
> 2.) In the TestWriteJokeConsoleOutput() test you guys didn't use the generics version of Assert.AreEqual (Assert.AreEqual-T). I would recommend using Assert.AreEqual-string since it appears you are trying to compare the string intended to be output and the string that the Jester class is outputting. (The "T" and "string" should be in angled brackets but for some reason they disappear when I post my comment)

Added generic to tests and implemented mock for properly testing the Chuck Norris filter.
Split the null test into two tests, hopefully covers enough.

File: CanHazFunny/CanHazFunny/IJokeOutput.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CanHazFunny;
+
+public interface IJokeOutput
+{
+    public void WriteJoke(string joke);
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/IJokeOutput.cs:12]:
```suggestion
    void WriteJoke(string joke);
```

File: CanHazFunny/CanHazFunny/IJokeService.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CanHazFunny;
+
+public interface IJokeService
+{
+    public string GetJoke();
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/IJokeService.cs:11]:
```suggestion
    string GetJoke();
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CanHazFunny;
+
+public class Jester : IJokeOutput, IJokeService
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:9]:
```suggestion
public class Jester : IJokeOutput, IJokeService
```
these should be implemented on their own classes, not here.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        JokeService = jokeService;
+    }
+
+    
+    public string TellJoke()
+    {
+        string testJoke = GetJoke();
+        bool isChuckNorris = true;
+        while (isChuckNorris) {
+            if (testJoke.Contains("Chuck Norris") || testJoke.Contains("Chuck") || testJoke.Contains("Norris")) {
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:49]: I'd suggest moving this if check into your while loop and getting rid of your `isChuckNorris` boolean all together.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                continue;
+            }
+            else
+            {
+                isChuckNorris=false;
+            }
+        }
+
+        ReallyCoolJokeOutput.WriteJoke(testJoke);
+        return testJoke;
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:60]: You don't need to return a string here, you could make this method void. Just depends what the intention is, I'm guessing you did this to make unit testing easier.
